src/management-system-v2/lib/ability/caslAbility.ts

```diff
@@ -60,7 +67,72 @@ const conditions = {
       'and',
       false,
     ),
-} as const;
+  $property_has_to_be_child_of: (valueInCondition: string, _, tree) => {
+    if (!tree)
+      throw new Error(
+        'If you specify a subtree (key: hasToBeChildOf) for a condition, you have to build the ability with a tree',
+      );
+
+    return (resource: any) => {
+      if (!valueInCondition) return false;
+
+      // Folder permissions are also applied to the folder itself
+      if (
+        (resource.__caslSubjectType__ as ResourceType) === 'Folder' &&
+        resource.id === valueInCondition
+      )
+        return true;
+
+      // If the resource doesn't specify a perent we can't know where it should be in the tree.
+      // Although this would probably be an error, this should be caught by zod, so here
+      // we just return false implying that this rule doesn't apply.
+      if (!resource.parentId) return false;
+
+      let currentFolder = resource.parentId;
+      while (currentFolder) {
+        if (currentFolder === valueInCondition) {
+          return true;
+        }
+        currentFolder = tree[currentFolder];
+      }
```

Comment:
This could only happen if there was an inconsistency in our folder structure (something like a folder is a parent of its parent). Which should never be the case, as all folder operations that would lead to this are not allowed. Still, I think it wouldn't be a real performance hit, and it could be useful if somehow folders got corrupted.